[inductor] enable bf32 for mkldnn linear pointwise/binary in inductor

zhuhaozhe opened the pull request:
When `torch.backends.mkldnn.matmul.fp32_precision=='bf16'`, we also enabled mkldnn linear in inductor path and allow to run with bf16 computation data type.

Testplan:
```
python test/inductor/test_mkldnn_pattern_matcher.py -k test_linear_unary
python test/inductor/test_mkldnn_pattern_matcher.py -k test_linear_fp32
python test/inductor/test_mkldnn_pattern_matcher.py -k test_multi_linear_share_same_input
```





Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #157520
* __->__ #127294



cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @peterbell10 @yf225 @ColinPeppler @desertfire

zhuhaozhe committed:
Update

[ghstack-poisoned]

jgong5 commented on torch/_inductor/fx_passes/mkldnn_fusion.py:
```diff
@@ -783,6 +783,12 @@ def is_linear_add_bias(match):
             add_node = match.output_node()
             linear_node = add_node.args[0]
             weight_meta = linear_node.args[1].meta.get("val")
+            if not hasattr(add_node.args[1], "meta"):
+                # May add an "int"
+                # We meet this issue while enabling bf32 for test_linear_unary
+                # the hardsimoid case will add "3" here and cannot check meta
+                # TODO: Further investigate on this issue
```
Is it for the bias node or something else? Can you double check? Why didn't we get the problem in the past?

zhuhaozhe replied:
It should be another input node for "add".
We did not meet this in the past because we only enabled and tested lp for mkldnn fusion. For lp case, there will be a "to_dtype" inserted by autocast and will not match this pattern.

I am trying to find a case to satisfy this pattern by raising a runtime error if this function will return True.
https://github.com/pytorch/pytorch/pull/127597

zhuhaozhe replied:
Fixed in https://github.com/pytorch/pytorch/pull/127597

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe commented:
@pytorchbot rebase

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe commented:
@pytorchbot rebase

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe commented:
@pytorchbot rebase

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe commented:
@pytorchbot rebase

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe committed:
Update

[ghstack-poisoned]

zhuhaozhe committed:
Update on "[inductor] enable bf32 for mkldnn linear pointwise/binary in inductor"

When `torch.backends.mkldnn.matmul.fp32_precision=='bf16'`, we also enabled mkldnn linear in inductor path and allow to run with bf16 computation data type.

Testplan:
```
python test/inductor/test_mkldnn_pattern_matcher.py -k test_linear_unary
python test/inductor/test_mkldnn_pattern_matcher.py -k test_linear_fp32
python test/inductor/test_mkldnn_pattern_matcher.py -k test_multi_linear_share_same_input
```








cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j commented:
@pytorchbot merge

yanbing-j commented:
Hi @jansel , could you please help review this PR as well? Thanks!

This PR is the follow-up of https://github.com/pytorch/pytorch/pull/125888, which is to refine fp32 precision API and has been merged. This PR will allow BF32 for mkldnn linear, which is use BF16 as FP32 internal precision for calculation.

yanbing-j committed:
Update

[ghstack-poisoned]

yanbing-j commented:
@pytorchbot merge
